### PR TITLE
feat: Output similar commands or script command when the input command is not correct

### PR DESCRIPTION
**Pull request by Zheaoli**
Signed-off-by: Manjusaka <me@manjusaka.me>

## Pull Request Checklist

- [x] A news fragment is added in `news/` describing what is new.
- [x] Test cases added for changed code.

## Describe what you have changed in this PR.

Support #3270

### CODE REVIEW

**Comment by Zheaoli** [comment]
image

File: src/pdm/cli/utils.py [hunk 1]
```diff
@@ -156,6 +157,36 @@ def parse_known_args(self, args: Any = None, namespace: Any = None) -> Any:
...
+            results.append((name, ratio))
+    return [name for name, _ in sorted(results, key=lambda x: x[1], reverse=True)]
+
+
+class ArgumentParserSimilarComandUtil:
+    color = {
+        "red": "\033[91m",
+        "green": "\033[92m",
+        "yellow": "\033[93m",
+    }
```
Comment (frostming) [inline src/pdm/cli/utils.py]: You can use this utility to get colorized message with rich:
https://github.com/pdm-project/pdm/blob/1fd94bd5044ea10bbc814d92144a2d1a1bd663fd/src/pdm/termui.py#L58-L70

For example

```python
message = style("[success]This text is green[/]")
```

File: src/pdm/cli/utils.py [hunk 2]
```diff
@@ -156,6 +157,36 @@ def parse_known_args(self, args: Any = None, namespace: Any = None) -> Any:
...
+
+    @classmethod
+    def format_similar_command(cls, root_command: str, commands: list[str], script_commands: list[str]) -> str:
+        similar_commands = find_similar_text(root_command, commands)
+        similar_script_commands = find_similar_text(root_command, script_commands)
+        message = f"""{cls.color['red']}Command not found: {root_command}
+    {cls.color['green']}Did you mean one of these command?
+        {cls.color['green']}{similar_commands}
```
Comment (frostming) [inline src/pdm/cli/utils.py:173]: Should better render the choices as seprate lines, like:

```
Did you mean one of these command?
  - foo
  - bar
```

File: src/pdm/cli/utils.py [hunk 3]
```diff
@@ -156,6 +157,36 @@ def parse_known_args(self, args: Any = None, namespace: Any = None) -> Any:
...
+def find_similar_text(origin_name: str, target_names: list[str], ratio: float = 0.5) -> list[str]:
+    results = []
+    for name in target_names:
+        ratio = SequenceMatcher(None, origin_name, name).ratio()
+        if ratio > 0.4:
+            results.append((name, ratio))
+    return [name for name, _ in sorted(results, key=lambda x: x[1], reverse=True)]
+
+
+class ArgumentParserSimilarComandUtil:
```
Comment (frostming) [inline src/pdm/cli/utils.py:160]: Just a personal flavor thing, I don't like static-method-only classes. Especially when the color attribute is removed, we can just make `format_similar_command` into a plain function.

**Review by frostming — COMMENTED** [review]
Thank you very much, it is a good start point and we can investigate how to implement the same for subcommands.

File: src/pdm/cli/utils.py (hunk 3 above)
Comment (Zheaoli) [reply src/pdm/cli/utils.py:160]: LGTM

File: src/pdm/cli/utils.py (hunk 1 above)
Comment (Zheaoli) [reply src/pdm/cli/utils.py]: nice ,I will dupdate it

**Review by frostming — APPROVED** [review]
The UI looks great

**Pull request merged**